### PR TITLE
[WIP] C-API <--> Python API for paddle.framework

### PR DESCRIPTION
**Pull request by reyoung**
* Add new C-API for refactored Paddle, in `paddle/c`
* Expose that C-API to paddle.v2.framework.c.Variable and
  paddle.v2.framework.c.Scope
  * Use Cython to warp them and give a native Python API.
* Add Unit tests.
  * See `python/paddle/v2/framework/tests/test_scope.py` for usage.

### CODE REVIEW

File: paddle/c/private.h [hunk 1]
```diff
@@ -0,0 +1,16 @@
+#pragma once
+
+namespace paddle {
+namespace c {
+
+enum ObjectType {
+  SCOPE = 0,
+};
+
+#define PADDLE_HANDLE_HEADER ::paddle::c::ObjectType type_;
```
Comment (wangkuiyi) [inline paddle/c/private.h]: 这个文件是干啥呢？

File: paddle/c/scope.cc [hunk 2]
```diff
@@ -0,0 +1,32 @@
+#include "scope.h"
```
Comment (wangkuiyi) [inline paddle/c/scope.cc]: If we use Boost::Python, all the following files in this PR

- scope.h 
- scope.cc
- private.h
- error.h
- dllexports.h

would be replaced by the following few-line C++ code:

```cpp
#include <boost/python.hpp>

BOOST_PYTHON_MODULE(pypaddle) {
    boost::python::class_<paddle::framework::Scope>("Scope")
        .def("create_var", &Scope::CreateVar)
        .def("create_scope", &Scope::CreateScope);
}
```

Please re-think about this necessity of create a C-API.

Comment (gangliao) [inline paddle/c/scope.cc]: Pybind11 is also an excellent candidate https://media.readthedocs.org/pdf/pybind11/master/pybind11.pdf

Caffe2 chose pybind11 to expose `c++11` directly.

Comment (reyoung) [inline paddle/c/scope.cc]: Whether uses C-API depends on is there any other languages need invoke Paddle C++ Core or not.

I am not sure only Python API is enough or not. At least there are several needs for us to give a C-API.

* Android/IOS API. Give a Java/Objective-C wrapper for Paddle.
* Go API. Needs for our Parameter Server.

Comment (reyoung) [inline paddle/c/scope.cc]: This PR is just an exploring PR for how to expose Paddle C++ core by C-API & Cython. I will give another PR by using pybind11.

**Pull request closed without merging**